R/write.R

```diff
@@ -0,0 +1,481 @@
+                x = supertbl_meta_labels, colNames = FALSE)
+  }
+
+  # Define redcap_data variable labels
+  var_labels <- supertbl$redcap_data %>% map(\(x) generate_dictionaries(x))
+
+  for (i in seq_along(supertbl$redcap_form_name)) {
+    wb$add_data(
+      sheet = sheet_vals[i],
+      x = var_labels[[i]], colNames = FALSE)
+  }
+
+  for (i in seq_len(nrow(wb$tables))) {
+    dims <- gsub("[0-9]+", "1", wb$tables$tab_ref[i])
+
+    wb$add_cell_style(sheet = i,
+                      dims = dims,
+                      wrapText = "1")
+    wb$add_font(sheet = i,
+                dims = dims,
+                color = openxlsx2::wb_color(hex = "7F7F7F"),
+                italic = "1")
+  }
+}
+
+#' @title Add the supertbl table of contents sheet
+#'
+#' @description
+#' Internal helper function. Adds appropriate elements to wb object. Returns
+#' a dataframe.
+#'
+#' @param supertbl a supertibble generated using `read_redcap()`
+#' @param wb An `openxlsx2` workbook object
+#' @param include_metadata Include a sheet capturing the combined output of the
+#' supertibble `redcap_metadata`. Default `TRUE`.
+#' @param labelled Whether or not to include labelled outputs, default `FALSE`.
+#' Requires use of `make_labelled()`.
+#' @param table_style Any excel table style name or "none" (see "formatting"
+#' vignette in \link[openxlsx2]{wb_add_data_table}). Default "tableStyleLight8".
+#' @param set_col_widths Width to set columns across the workbook. Default
+#' "auto", otherwise a numeric value. Standard Excel is 8.43.
+#'
+#' @importFrom dplyr select mutate row_number across
+#' @importFrom tidyselect any_of
+#'
+#' @returns A dataframe
+#'
+#' @keywords internal
+
+add_supertbl_toc <- function(wb,
+                             supertbl,
+                             include_metadata,
+                             labelled,
+                             table_style,
+                             set_col_widths) {
+
+  # To avoid XLSX indicators of "Number stored as text", change class type
+  convert_percent <- function(x) {
+    class(x) <- c("numeric", "percentage")
+    x
+  }
+
+  supertbl_toc <- supertbl %>%
+    # Remove list elements
+    select(-any_of(c("redcap_data", "redcap_metadata", "redcap_events"))) %>%
+    # Necessary to avoid "Number stored as text" Excel dialogue warnings
+    mutate(across(any_of("data_na_pct"), convert_percent))
+
+  # Conditionally Add metadata default to TOC
+  if (include_metadata) {
+    supertbl_toc[nrow(supertbl_toc) + 1, 1] <- "REDCap Metadata"
+  }
+
+  # Re-establish lost label(s) by referencing original labels and indexing
+  # Generalized for future proofing
+  labelled::var_label(supertbl_toc) <- labelled::var_label(supertbl)[names(labelled::var_label(supertbl_toc))]
+
+  # Add custom Sheet # column and label
+  supertbl_toc <- supertbl_toc %>%
+    mutate(`Sheet #` = row_number())
+
+  # Assign label
+  if (labelled) {
+    labelled::var_label(supertbl_toc)$`Sheet #` <- "Sheet #"
+  }
```

Comment:
Currently `labelled == TRUE` doesn't guarantee `labelled` is installed

Reply:
Hm, instead of adding `check_installed` to a few places, maybe it's better to include `check_installed` during the designation of `labelled`

Reply:
Yeah I think if you can put it in `check_labelled()` and have `check_labelled()` only return `TRUE` if `check_installed("labelled")` passes then downstream we can safely use `labelled::function()` as long as it's behind an `if (labelled) { ...` conditional.

Reply:
Do you recall if wound up resolving this? I think we did?

Reply:
Yeah this was resolved. `check_labelled()` now only returns `TRUE` if `labelled` is installed.